docs/docs/forms.mdx

```diff
@@ -360,74 +360,104 @@ to fully customize the validation process.
+   custom way.
+2. Make sure that in the domain file you list for your form only those slots that don't
+   use custom mappings.
+3. Override `required_slots` to add all slots with custom mappings to the list of slots
+   the form should request.
+
+The following example shows the implementation of a form which extracts a slot
+`outdoor_seating` in a custom way, in addition to the slots whose mappings are defined
+in the domain.
+The method `extract_outdoor_seating` sets the slot `outdoor_seating` based on whether
+the keyword `outdoor` was present in the last user message.
 
 ```python
-from typing import Dict, Text, List
+from typing import Dict, Text, List, Optional, Any
 
 from rasa_sdk import Tracker
-from rasa_sdk.events import EventType
 from rasa_sdk.executor import CollectingDispatcher
-from rasa_sdk import Action
-from rasa_sdk.events import SlotSet
+from rasa_sdk.forms import FormValidationAction
 
 
-class ValidateRestaurantForm(Action):
+class ValidateRestaurantForm(FormValidationAction):
     def name(self) -> Text:
         return "validate_restaurant_form"
 
-    def run(
+    async def required_slots(
+        self,
+        slots_mapped_in_domain: List[Text],
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> Optional[List[Text]]:
+        required_slots = slots_mapped_in_domain + ["outdoor_seating"]
+        return required_slots
+
+    async def extract_outdoor_seating(
         self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
-    ) -> List[EventType]:
+    ) -> Dict[Text, Any]:
         text_of_last_user_message = tracker.latest_message.get("text")
+        sit_outside = "outdoor" in text_of_last_user_message
 
-        return [SlotSet("user_message_length", len(text_of_last_user_message))]
+        return {"outdoor_seating": sit_outside}
 ```
 
-### Requesting Extra Slots
+By default the `FormValidationAction`
+will automatically set the [`requested_slot`](forms.mdx#the-requested_slot-slot) to the
+first slot specified in `required_slots` which is not filled.
 
-If you make frequent changes to the required slots and don't want to retrain your
-assistant when your form changes, you can also use a
-[Custom Action](./actions.mdx#custom-actions) `validate_{form_name}` to define
-which slots should be requested. Rasa Open Source will run your custom action whenever
-the form validates user input. Set the slot `requested_slot` to the name of the slot
-which should be extracted next. If all desired slots are filled, set `requested_slot`
-to `None`.
+### Dynamic Form Behavior
 
-The following example shows the implementation of a custom action which requests
-the three slots `cuisine`, `num_people`, and `outdoor_seating`.
+By default Rasa Open Source will ask for the next
+[`requested_slot`](forms.mdx#the-requested_slot-slot) defined in the slot mappings
+in the domain, which is not filled yet. If you use
```

Comment:
Hm, I'm trying to think of a better way to refer to the slots listed for the form in the domain. How about:
```suggestion
By default Rasa Open Source will ask for the next empty slot from the slots
listed for your form in the domain file. If you use
```